apollo-router/src/plugins/expose_query_plan.rs

```diff
@@ -0,0 +1,241 @@
+use futures::future::ready;
+use futures::stream::once;
+use futures::StreamExt;
+use http::HeaderValue;
+use serde_json_bytes::json;
+use tower::util::BoxService;
+use tower::BoxError;
+use tower::ServiceExt as TowerServiceExt;
+
+use crate::layers::ServiceExt;
+use crate::plugin::Plugin;
+use crate::plugin::PluginInit;
+use crate::register_plugin;
+use crate::services::QueryPlannerContent;
+use crate::services::QueryPlannerRequest;
+use crate::services::QueryPlannerResponse;
+use crate::services::RouterRequest;
+use crate::services::RouterResponse;
+
+const EXPOSE_QUERY_PLAN_HEADER_NAME: &str = "Apollo-Expose-Query-Plan";
+const ENABLE_EXPOSE_QUERY_PLAN_ENV: &str = "APOLLO_EXPOSE_QUERY_PLAN";
+const QUERY_PLAN_CONTEXT_KEY: &str = "experimental::expose_query_plan.plan";
+const ENABLED_CONTEXT_KEY: &str = "experimental::expose_query_plan.enabled";
+
+#[derive(Debug, Clone)]
+struct ExposeQueryPlan {
+    enabled: bool,
+}
+
+#[async_trait::async_trait]
+impl Plugin for ExposeQueryPlan {
+    type Config = bool;
+
+    async fn new(init: PluginInit<Self::Config>) -> Result<Self, BoxError> {
+        Ok(ExposeQueryPlan {
+            enabled: init.config,
+        })
+    }
+
+    fn query_planning_service(
+        &self,
+        service: BoxService<QueryPlannerRequest, QueryPlannerResponse, BoxError>,
+    ) -> BoxService<QueryPlannerRequest, QueryPlannerResponse, BoxError> {
+        service
+            .map_response(move |res| {
+                if res
+                    .context
+                    .get::<_, bool>(ENABLED_CONTEXT_KEY)
+                    .ok()
+                    .flatten()
+                    .is_some()
+                {
+                    if let QueryPlannerContent::Plan { plan, .. } = &res.content {
+                        res.context
+                            .insert(QUERY_PLAN_CONTEXT_KEY, plan.root.clone())
+                            .unwrap();
+                    }
+                }
+
+                res
+            })
+            .boxed()
+    }
+
+    fn router_service(
+        &self,
+        service: BoxService<RouterRequest, RouterResponse, BoxError>,
+    ) -> BoxService<RouterRequest, RouterResponse, BoxError> {
+        let conf_enabled = self.enabled;
+        service
+            .map_future_with_context(move |req: &RouterRequest| {
+                let is_enabled = (conf_enabled || std::env::var(ENABLE_EXPOSE_QUERY_PLAN_ENV).as_deref() == Ok("true")) && req.originating_request.headers().get(EXPOSE_QUERY_PLAN_HEADER_NAME) == Some(&HeaderValue::from_static("true"));
```

Comment:
could we avoid checking the env var on every call? This has to go through locks and allocations: https://github.com/rust-lang/rust/blob/e50ff9b4521234e56ff46f8ed0372d5cb5689654/library/std/src/sys/unix/os.rs#L531